### PR TITLE
ASIM parsers filtering validations - dynamic tests - part1

### PR DESCRIPTION
**Pull request by TomerAligaev**
- Add utility functions for the dynamic parameters tests as part of the ASim parsers filtering validations.
- Rename the tests script file.
- Add Readme file.

### CODE REVIEW

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py [hunk 1]
```diff
@@ -77,6 +77,12 @@ def create_parameters_string(parser_file):
     return ','.join(paramsList)
 
 
+# Creating a string of the values in the list with colons between them
```
Comment (ori-licht) [inline ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:80]: >colons [](http://example.com/codeflow?start=51&length=6)

worth adding an example, and is that colons? or commas?

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py [hunk 2]
```diff
@@ -94,6 +100,55 @@ def create_execution_strings_with_one_parameter(parameter, value, column_name):
...
+        return str
+    # Checking if there is at least one value in the query response that the prefix is not contained in.
+    for row in query_response.tables[0].rows:
+        value = row[0]
+        if substring not in value:
+            return substring
+    return str
+
+
+def get_postfix(str, query_response, current_list):
```
Comment (ori-licht) [inline ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:128]: >postfix [](http://example.com/codeflow?start=8&length=7)

?
maybe suffix?

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py [hunk 3]
```diff
@@ -218,6 +273,105 @@ def scalar_test_check_fictive_value(self, param_name, query_definition, column_n
             self.assertEqual(0, len(no_results_response.tables[0].rows), f"Parameter: {param_name} - Returned results for non existing filter value. Filtered by value: {DUMMY_VALUE}")
 
         
+    # Returning an array of at most two values from the query response, each string in the returned array is not contained in the other
```
Comment (ori-licht) [inline ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:276]: this comment is not clear, what is the meaning of "each string in the returned array is not contained in the other"

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py [hunk 4]
```diff
@@ -218,6 +273,105 @@ def scalar_test_check_fictive_value(self, param_name, query_definition, column_n
...
+            if value == "":
+                continue
+            for row2 in query_response.tables[0].rows:
+                value2 = row2[0]
+                if value not in value2:
+                    values.append(value)
+                    break
+            if len(values) == 2:
+                break
+        return values
```
Comment (ori-licht) [inline ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:302]: let's talk about this, it is not understood why you are dealing with the case where the length is 1, 2. 
why the general approach is not good enough

Comment (TomerAligaev) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:302]: For the case of length 2 you are correct. I will remove it.
For the case of length 1 lets assume we have a response with only one value. The if condition in line 297 (if value not in value2) will not be true as both "value" and "value2" will have the value of the only element in the response.
As a result, this one value won't be appended to the returned array even though it should be appended.
By the way, following our teams meeting, this is the same reason why the case of length 1 is checked in both get_prefix and get_postfix.

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py [hunk 5]
```diff
@@ -218,6 +273,105 @@ def scalar_test_check_fictive_value(self, param_name, query_definition, column_n
...
+
+    def add_substring_to_list(self, query_response, substrings_list, num_of_substrings):
+        '''
+        The function adds at most "num_of_substrings" substrings of values from "query_response" to "substrings_list".
+        A substring of a value will be either its postfix from after the first dot in the value or its prefix until the first dot in the value.
+        '''
+
+        # Looking for values with substrings that can be appended to the list
+        for row in query_response.tables[0].rows:
+            value = row[0]
```
[the remark this answers is not in the document]
Comment (TomerAligaev) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:332]: The rows in query_response are a result of a KQL query summarized and counted by a certain row.
For example:
_Im_NetworkSession_Microsoft365DefenderV04() | summarize count() by IpAddr 

image

So the value relevant to each row will be in row[0].

**Comment by ori-licht** [comment]
if the tests are for query_response.tables[0].rows[0] in all cases, why not just pass that to all functions?
you are constantly using the "query_response" and inside the function you always have:
"query_response.tables[0].rows[0]"

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py (hunk 4 above)
Comment (ori-licht) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:302]: Let's talk

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ori-licht) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:332]: not sure I understand the last sentence:
"the value relevant to each row will be in row[0]."

what do you mean by "each row" if you are only looking at row[0]?

**Comment by v-atulyadav** [comment]
Hi @TomerAligaev, please act on @ori-licht's comments. Thanks

**Comment by v-atulyadav** [comment]
Hi @TomerAligaev, waiting for your response on @ori-licht's comments. Thanks

**Comment by ori-licht** [comment]
> Hi @TomerAligaev, waiting for your response on @ori-licht's comments. Thanks

Hi @v-atulyadav , Tomer is OOF for the upcoming weeks.

**Comment by v-atulyadav** [comment]
sure @ori-licht thank you for this update. We look forward to Tomer's response.

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py (hunk 4 above)
Comment (ori-licht) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:302]: def get_values_for_dynamic_tests(self, rows):
	if len(rows) == 1:
		return [] if rows[0][0] == "" else return [rows[0][0]]
		
	values = []
	# Searching values in rows which are not contained in at least one other value
	for row in rows:
		value = row[0]
		# if the value in an empty string we want to skip it
		if value == "":
			continue
			
		# if we already found 2 values that sutisfy the condifitons we can return them 
		if len(values) == 2:
			break
			
		for rowToCompareWith in rows:
			valueToCompareWith = rowToCompareWith[0]
			if value not in valueToCompareWith:
				values.append(value)
				# if we found one value that sutisfy the condition, we can continue to the next one
				break
	return values

**Comment by TomerAligaev** [comment]
Passing now only rows instead of query_response.

---
In reply to: [1618014737](https://github.com/Azure/Azure-Sentinel/pull/8408#issuecomment-1618014737) [](http://example.com/codeflow?ancestors=1618014737)

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py (hunk 1 above)
Comment (TomerAligaev) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py]: Indeed it is commas, my mistake.
And added example.

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py (hunk 3 above)
Comment (TomerAligaev) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py]: Changed the comment.

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TomerAligaev) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:341]: Changed the function after talking face to face.

File: ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py (hunk 4 above)
Comment (TomerAligaev) [reply ASIM/dev/ASimTester/filteringTest/ASimFilteringTest.py:308]: Changed the function after talking face to face.

**Review by ori-licht — APPROVED** [review]

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**